pkg/kubelet/kubelet_pods.go

```diff
@@ -1635,7 +1635,18 @@ func (kl *Kubelet) generateAPIPodStatus(pod *v1.Pod, podStatus *kubecontainer.Po
 				}
 				// Secondary IP is not set #105320
 				if len(hostIPs) == 2 && len(s.PodIPs) == 1 {
-					s.PodIPs = append(s.PodIPs, v1.PodIP{IP: hostIPs[1].String()})
+					ok, err := utilnet.IsDualStackIPStrings([]string{s.PodIPs[0].IP, hostIPs[1].String()})
+					if err != nil {
+						kl.recorder.Eventf(pod, v1.EventTypeWarning, "InvalidHostIPs", "host IP is invalid %q", hostIPs[1].String())
+					} else {
+						if ok {
+							s.PodIPs = append(s.PodIPs, v1.PodIP{IP: hostIPs[1].String()})
+						} else {
+							kl.recorder.Eventf(pod, v1.EventTypeWarning, "InvalidHostIPs",
+								"Inconsistent addresses order received from the cloud-provider, IP family of the second host IP %q is the same as the first one configured in the kubelet %q",
```

Comment:
The wording here is not great in terms of helping the admin understand what to do.

Also, if we are going to notice the "you requested IPv4 but the cloud provider wants dual-stack IPv6-primary" case, then we should really also notice the "you requested IPv4 but the cloud provider wants _single-stack_ IPv6" case too. (It is likely that the original reporter was also running into that case but just didn't notice yet because it doesn't cause any explicit errors. But it may cause problems later with Services.)

So, starting at line 1627:
- if `utilnet.IPFamilyOfString(s.HostIP) != utilnet.IPFamilyOf(hostIPs[0])` then we need to log a warning:
  - if `s.HostIP` is IPv4 then the problem is that kubelet auto-detected an IPv4 node IP, but the cloud provider wants this to be a single-stack IPv6 or dual-stack IPv6-primary node, and so the fix is for the admin to either (a) pass `--node-ip ::` to kubelet to tell it to prefer IPv6 like the cloud provider wants, (b) pass `--node-ip <EXPLICIT-IPV4-ADDRESS>` to request a single-stack IPv4 node with a specific IP. (Once `CloudDualStackNodeIPs` is beta we can also suggest (c) pass `--node-ip <EXPLICIT-IPV4-ADDRESS>,<EXPLICIT-IPV6-ADDRESS>`.)
  - if `s.HostIP` is IPv6 then the problem is that the admin passed `--node-ip ::` to request an IPv6 node IP, but the cloud provider wants this to be a single-stack IPv4 or dual-stack IPv4-primary node, and so the fix is for the admin to either (a) _not_ pass `--node-ip ::` to kubelet and let it default to IPv4-primary like the cloud provider wants, (b)/(c) pass an explicit IPv6 IP as above
- regardless of whether it logged a warning, it should do `s.HostIP = hostIPs[0].String()`, as it does now
- if the pod is host-network then:
  - it should set `s.PodIPs[0]` to `hostIPs[0]` if `s.PodIPs` is unset, as it does now
  - it should set `s.PodIPs[1]` to `hostIPs[1]` if `len(hostIPs) == 2 && len(s.PodIPs) == 1` _and_ `utilnet.IPFamilyOfString(s.PodIPs[0].IP) != utilnet.IPFamilyOf(hostIPs[1])`
    - (the family mismatch can only happen if we have already logged an error above, so we don't need to warn again here)